### PR TITLE
Feature: Chat Title Generation

### PR DESCRIPTION
**Pull request by alashchev17**
# Feature: Chat Title Generation

## Description

- Previously we were taking as a chat title first user message, but in reality we are not always making self-explanatory requests and even not always doing requests as a text. Sometimes we just send some code blocks and have them appearing as a chat title is not what we need.
- Implemented some logic across Redux Async Thunks to make a request to generate a chat title based on first 2 messages (User & Assistant) from LSP.
- [Any background context or related links?](https://github.com/orgs/smallcloudai/projects/5/?pane=issue&itemId=78690431)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, only code improvements)
- [ ] Documentation update

## How to Test

- Step 1: Open a New Chat
- Step 2: Start asking a chat for some assistance (anything)
- Step 3: Get the nice and short chat title!

### Functionality Screencast
Functionality Screencast

## Checklist

- [x] My code follows the code style of this project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I have updated the documentation where necessary.

### CODE REVIEW

File: src/features/History/historySlice.ts [hunk 1]
```diff
@@ -85,6 +80,13 @@ export const historySlice = createSlice({
         b.updatedAt.localeCompare(a.updatedAt),
       ),
   },
+  extraReducers(builder) {
+    builder.addCase(chatGenerateTitleThunk.fulfilled, (state, action) => {
+      // making a logic on fulfilled status of thunk and updating title of a chat thread
+      if (!isChatGetTitleActionPayload(action.payload)) return state;
+      state[action.payload.chatId].title = action.payload.title;
```
Comment (MarcMcIntosh) [inline src/features/History/historySlice.ts:87]: Chat might be in he cache

Comment (MarcMcIntosh) [reply src/features/History/historySlice.ts:87]: does it save the chat after adding the title?

Comment (MarcMcIntosh) [reply src/features/History/historySlice.ts:87]: oh wait this is the history, ...so it should work :)

**Review by MarcMcIntosh — COMMENTED** [review]
I'll give this a few runs, but it looks like it should work :)

**Comment by MarcMcIntosh** [comment]
Note: small issue where stream fails causes the title not to be made (probably due to an uncleared error message).

**Review by MarcMcIntosh — CHANGES_REQUESTED** [review]
Looks good, there's just one small issue when the assistant message has tool calls that haven't been called.

**Comment by olegklimov** [comment]
1. Titles are too generic (should be specific about files, symbols, objects, etc)
2. Only give it the first user message and nothing else
3. Detect stop condition reliably
4. title() function in each chat that either returns calculated title, or first 50 characters for the first user question

**Comment by alashchev17** [comment]
The only thing that remains is to review a prompt for chat title generation. Probably, it can be written better.

**Review by MarcMcIntosh — APPROVED** [review]
LGTM

**Pull request merged**